### PR TITLE
XGM

### PR DESCRIPTION
**Pull request by PJB3005**
So last year in August, me and @clusterfack started this. We lost steam and forgot about it. To not lose steam this time I'm just gonna make the PR so people can shout when.

I'm reviving this now, but it's still very WiP and in no way merge ready.

I lost the commit history due a Git issue, see here for the history: https://github.com/vgstation-coders/vgstation13/tree/XGM

Checklist:
* [ ] `#warn` TODOs need to be fixed.
* [ ] Zone regen has bug where sprites get lost for a tick.
* [ ] Fire needs balancing.
* [ ] Update with latest fixes from bay.
* [ ] Make plasma not ignite below visibility.

### CODE REVIEW

**Comment by JustSumBody** [comment]
remind me what all XGM does special?
I remember there's the whole "fires melt walls" dealio and MULTIZAS, but other than that what does it do

**Comment by PJB3005** [comment]
Actually this is too old for multiZAS but porting MultiZAS will become easier at least.

**Comment by ghost** [comment]
holy
fuck


also what is multizas

**Comment by Anton-Kr** [comment]
@unusualcrow Working ZAS on multi-z stations.

**Comment by PJB3005** [comment]
Nevermind actually I think this does include multiZAS somewhat.

**Comment by Exxion** [comment]
Even our current ZAS includes multiZAS
There's a define to enable it
`#define ZLEVELS`

It almost definitely doesn't work properly, but it's there

**Comment by FudgePucker** [comment]
I guess multi-zas could lead to some away mission fun.

What other possible changes would this system provide?

**Comment by PJB3005** [comment]
Bugfixes, better performance, meme gases with basically zero cost, fun fires, Wall melting.

**Comment by gbasood** [comment]
Gas datums

**Comment by Fruktlimpa** [comment]
Wall melting. At what temperature are we talking here and how fast will the walls melt? I can imagine it could get overwhelming.

**Comment by PJB3005** [comment]
No idea fight now and it's probably gonna need balancing for engineering.

**Comment by Fruktlimpa** [comment]
Sounds really neat in any case.

**Comment by Shadowmech88** [comment]
What exactly is this? A new atmos system?

What all does it do differently?

**Comment by PJB3005** [comment]
It's just updating everything to (almost) latest bay.

**Comment by Shadowmech88** [comment]
So what all will it do differently?
I know nothing of current Bay atmos.

**Comment by Exxion** [comment]
For one, datumized gases
I haven't read the code, but I assume they're singleton datums defining *types* of gases, rather than an individual datum for each instance of a gas in a zone, for performance reasons
The latter... didn't work very well when we tried it

*shudder

**Comment by PJB3005** [comment]
Yeah they're just singletons.

**Comment by ComicIronic** [comment]
We never tried new datum instances for every gas instances. The previous version also used singletons - it performed badly for different reasons.

**Comment by Exxion** [comment]
Oh
Well never mind

**Comment by clusterfack** [comment]
We had a ton of checks and recalculations for whenever any type of gas was added/removed, and since about 1000 gas datums are created per second iirc it was a problem.

File: __DEFINES/atmos.dm [hunk 1]
```diff
@@ -0,0 +1,102 @@
+#define CELL_VOLUME        2500 // Liters in a cell.
+#define MOLES_CELLSTANDARD (ONE_ATMOSPHERE * CELL_VOLUME / (T20C * R_IDEAL_GAS_EQUATION)) // Moles in a 2.5 m^3 cell at 101.325 kPa and 20 C.
+#define MOLES_CELLMARS (MARS_ATMOSPHERE*CELL_VOLUME/(T20C*R_IDEAL_GAS_EQUATION)) //Same as above but for mars (temperature is 20 degrees - it's assumed that it's noon on Mars)
+
+#define O2STANDARD 0.21 // Percentage.
+#define N2STANDARD 0.79
+
+#define MARS_ATMOSPHERE		0.6 //kPa
+#define CO2MARS 0.96
+#define N2MARS  0.04 //Mars atmosphere is actually 1.9% nitrogen, 1.9% argon with traces of other gases. Simplified to 4% nitrogen
```
Comment (Probe1) [inline __DEFINES/atmos.dm:10]: Why do we have separate defines for Mars?  (This goes for all the defines but I'm not going to clutter up every line below)

File: __DEFINES/global.dm [hunk 2]
```diff
@@ -168,11 +168,8 @@ var/polarstar = 0 //1 means that the polar star has been found, 2 means that the
 // nanomanager, the manager for Nano UIs
 var/datum/nanomanager/nanomanager = new()
 
-#define SPEED_OF_LIGHT 3e8 //not exact but hey!
```
Comment (Probe1) [inline __DEFINES/global.dm:171]: Why was this even defined?

File: __DEFINES/atmos.dm (hunk 1 above)
Comment (unid15) [reply __DEFINES/atmos.dm:10]: Some turfs use mars atmosphere

File: __DEFINES/global.dm (hunk 2 above)
Comment (PJB3005) [reply __DEFINES/global.dm:171]: FYI it still is I just moved it.

**Comment by PJB3005** [comment]
I am already in love: http://84.195.252.227/static/ShareX/2017/%25m/2017-04-15_16-36-21.mp4

File: __DEFINES/atmos.dm (hunk 1 above)
Comment (ghost) [reply __DEFINES/atmos.dm:10]: one of the vaults uses mars atmosphere

**Comment by mtpowell** [comment]
This looks amazing, how does the TEG burn room hold up and the SME? Does it act on pressure?

**Comment by PJB3005** [comment]
Wall melting and window breaking only happens from fires, probably for performance. Melting temperature of reinforced walls is literally infinite and I'm gonna make plasma glass not break either.

I want to remind you that windows not breaking previously from fire was a bug.

**Comment by Exxion** [comment]
Windows *do* break from pressure
But only from the pressure on their own tile, rather than the differential across them

**Comment by PJB3005** [comment]
@Exxion pretty sure they only break from fire, and XGM fixes the same tile issue with it.

**Comment by Exxion** [comment]
Let me rephrase that
When they break, it's due to a pressure check
I don't remember whether or not that check is only performed if there's fire there, though

**Comment by Shadowmech88** [comment]
Is there any kind of remotely realistic reason why a material would only melt when exposed to flame rather than temperature?
Or are we just going to have to ignore the inconsistency?

**Comment by Exxion** [comment]
Performance

**Comment by Shadowmech88** [comment]
I get why it's necessary but when possible I still like to have even thinly-constructed in-universe reasons for why it could happen.

**Comment by Exxion** [comment]
Oxidation

**Comment by BlueNexus** [comment]
You'll need to port the MultiZAS changes as well, your version is way too old to work properly.

**Comment by PJB3005** [comment]
I couldn't find any significant changes to ZAS since when we started this, though I'll look again.

**Comment by BlueNexus** [comment]
https://github.com/Baystation12/Baystation12/pull/14538
https://github.com/Baystation12/Baystation12/pull/14558
https://github.com/Baystation12/Baystation12/pull/14927
https://github.com/Baystation12/Baystation12/pull/14957
@PJB3005 You'll need all of these for MultiZAS to work properly. Comment or uncomment the #define MultiZAS flag at your discretion.

**Comment by PJB3005** [comment]
@BlueNexus alright, I'll port em, thanks!

**Comment by BlueNexus** [comment]
Oh, this too - https://github.com/Baystation12/Baystation12/pull/14225
Port that one first. Sorry, I forgot about that one. (And, no problem. 😄  )

**Comment by BobdaBiscuit** [comment]
RIP

**Comment by BlueNexus** [comment]
RIP

Why, if I may ask?

**Comment by Exxion** [comment]
Not again

**Comment by gbasood** [comment]
drama
possibly revived later

**Comment by Exxion** [comment]
I'm going to set literally everything that has ever existed or will ever exist on fire
Why does everything go to shit roughly once a month?

**Comment by PJB3005** [comment]
@Exxion I think you know the answer.

**Comment by Exxion** [comment]
>the

There isn't only one

**Comment by FudgePucker** [comment]
I don't know what happened so I will just guess what happened.
Discord was a mistake.

**Comment by gbasood** [comment]
wrong
